### PR TITLE
Feature/monitor

### PR DESCRIPTION
**Pull request by szkiba**
Add /v1/monitor endpoint to serve metrics in Prometheus text format.

See  #1787

### CODE REVIEW

**Review by MStoykov — COMMENTED** [review]
HI @szkiba, sorry for the slow response and the lint problems :(. We are nearing the v0.30.0 release so we have some stuff we need to finish up, and also catching up after the holidays :).

Let's start with that I am somewhat of the opinion that this should be an output (extension) ... which just has an endpoint (possibly on another port that isn't the REST API). Unfortunately, we currently don't have output extensions, although hopefully work on it will start in the v0.31 cycle, no promises though ;) And I don't know if you can get the Engine from an output, but I am also not certain it is such a good idea to do it that way(see my comment on races).

Your use case is interesting but IMO for this to be in k6 it should be possible to give you all the metrics split by each tag that has been emitted, which is a much bigger change ... including a lot of other things that need to be changed inside of k6 if not for any other reasons but for performance.  This is closely related to #1321 and likely fixing that will fix my issue with this PR :shrug: 

The major and obvious blocker though is that you updated way too many dependencies IMO and as #1758 shows this might not be ... a good idea ;). Can you change the PR to only add the stuff you need instead of updating stuff as sarama, the kafka library we use for another output?

The above doesn't mean that this will get or not get merged based on any of the changes that I have proposed ;). This will need more reviews and figuring out if this change (in any of its forms) will be useful given everything else that is planned in k6, and given the fact that we are very likely (IMO) to change significantly how metrics are handled internally  - either making this change one more thing to refactor or something that will be much easier to do.

File: api/v1/monitor_routes.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	if engine.ExecutionScheduler != nil {
+		t = engine.ExecutionScheduler.GetState().GetCurrentTestRunDuration()
+	}
+
+	metrics := make([]dto.MetricFamily, 0)
+	for _, m := range engine.Metrics {
+		metrics = append(metrics, newMetricFamily(m, t)...)
+	}
+
+	data, err := marshallMetricFamily(metrics)
```
[the remark this answers is not in the document]
Comment (szkiba) [reply api/v1/monitor_routes.go:46]: Thank you, I just made it similar to metrics endpoint and didn't spent too much time on it...

File: api/v1/monitor.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		metrics = append(metrics, newGauge(m.Name+"_max", trend.Max, m.Name+" maximum value"))
+		metrics = append(metrics, newGauge(m.Name+"_avg", trend.Avg, m.Name+" average value"))
+		metrics = append(metrics, newGauge(m.Name+"_med", trend.Med, m.Name+" median value"))
+		metrics = append(metrics, newGauge(m.Name+"_p90", trend.P(0.90), m.Name+" 90 percentile value"))
+		metrics = append(metrics, newGauge(m.Name+"_p95", trend.P(0.95), m.Name+" 95 percentile value"))
+	case stats.Rate:
+		rate := m.Sink.(*stats.RateSink)
+		metrics = append(metrics, newGauge(m.Name+"_rate", float64(rate.Trues)/float64(rate.Total),
+			m.Name+" percentage of non-zero values"))
+	}
```
[the remark this answers is not in the document]
Comment (szkiba) [reply api/v1/monitor.go:57]: Probably counter fits better

**Comment by szkiba** [comment]
> HI @szkiba, sorry for the slow response and the lint problems :(. We are nearing the v0.30.0 release so we have some stuff we need to finish up, and also catching up after the holidays :).
> 
> Let's start with that I am somewhat of the opinion that this should be an output (extension) ... which just has an endpoint (possibly on another port that isn't the REST API). Unfortunately, we currently don't have output extensions, although hopefully work on it will start in the v0.31 cycle, no promises though ;) And I don't know if you can get the Engine from an output, but I am also not certain it is such a good idea to do it that way(see my comment on races).
> 
> Your use case is interesting but IMO for this to be in k6 it should be possible to give you all the metrics split by each tag that has been emitted, which is a much bigger change ... including a lot of other things that need to be changed inside of k6 if not for any other reasons but for performance. This is closely related to #1321 and likely fixing that will fix my issue with this PR 
> 
> The major and obvious blocker though is that you updated way too many dependencies IMO and as #1758 shows this might not be ... a good idea ;). Can you change the PR to only add the stuff you need instead of updating stuff as sarama, the kafka library we use for another output?
> 
> The above doesn't mean that this will get or not get merged based on any of the changes that I have proposed ;). This will need more reviews and figuring out if this change (in any of its forms) will be useful given everything else that is planned in k6, and given the fact that we are very likely (IMO) to change significantly how metrics are handled internally - either making this change one more thing to refactor or something that will be much easier to do.

Thank you, you review was useful. Probably the k6 extension (the xk6 thing) is a good direction. I didn't realized it, because its a new feature (I read blogs about one years ago). It seem very promising. An I'm happy to see the gRPC support, it is also new  for me. Congratulations, k6 become better and better!

**Comment by szkiba** [comment]
Hi @MStoykov , I got  a few hours free time to rewrite prometheus http exporter as xk6 extension. I found a cool new feature in k6 release notes: output extension. It is really cool, congrat to k6 team! Using this feature, I reimplemented the exporter as output extension and works fine. https://github.com/szkiba/xk6-prometheus
Thanx again.

**Comment by szkiba** [comment]
Prometheus HTTP exporter rewrote as xk6 output extension so this PR is obsolote now.
https://github.com/szkiba/xk6-prometheus

**Pull request closed without merging**